restli-tools/src/main/java/com/linkedin/restli/tools/data/ExtensionSchemaValidationCmdLineApp.java

```diff
@@ -179,48 +205,102 @@ private static void checkExtensionSchemaFields(List<RecordDataSchema.Field> exte
       {
         if (!(dataElement instanceof DataMap))
         {
-          _logger.error("Extension schema annotation is not a datamap!");
-          System.exit(1);
+          throw new InvalidExtensionSchemaException("Extension schema annotation is not a datamap!");
         }
         DataSchema extensionSchemaAnnotationSchema = new ExtensionSchemaAnnotation().schema();
         ValidationResult result = ValidateDataAgainstSchema.validate(dataElement, extensionSchemaAnnotationSchema, validationOptions);
         if (!result.isValid())
         {
-          _logger.error("Extension schema annotation is not valid: " + result.getMessages());
-          System.exit(1);
+          throw new InvalidExtensionSchemaException("Extension schema annotation is not valid: " + result.getMessages());
         }
       }
+      catch (InvalidExtensionSchemaException e)
+      {
+        throw e;
+      }
       catch (Exception e)
       {
         _logger.error("Error while checking extension schema field annotation: " + e.getMessage());
         System.exit(1);
       }
+      checkExtensionSchemaFieldSchema(field.getType(), properties);
     }
   }
 
-  private static void checkExtensionSchemaFieldSchema(DataSchema fieldSchema)
+  private static void checkExtensionSchemaFieldSchema(DataSchema fieldSchema, Map<String, Object> extensionAnnotations)
+      throws InvalidExtensionSchemaException
   {
-    if (!isAnnotatedWithResourceKey(fieldSchema) && !arrayAndItemIsAnnotatedWithResourceKey(fieldSchema))
+    checkExtensionSchemaFieldSchemaType(fieldSchema);
+    if (!isAnnotatedWithResourceKey(fieldSchema, extensionAnnotations) && !arrayAndItemIsAnnotatedWithResourceKey(fieldSchema, extensionAnnotations))
     {
-        _logger.error("Field schema: [{}] is not annotated with 'resourceKey'", fieldSchema.toString());
-        System.exit(1);
+       throw new InvalidExtensionSchemaException("Field schema: " + fieldSchema.toString() + " is not annotated with 'resourceKey'");
```

Comment:
This error message might be misleading for Arrays a it would say the Array schema is missing annotation, whereas the correct message is for the array item's type. 

I think you can inline the logic in checkExtensionSchemaFieldSchemaType here, and then just use the modified fieldSchema to call isAnnotatedWithResourceKey

You can delete arrayAndItemIsAnnotatedWithResourceKey

you should also consider in-lining isAnnotatedWithResourceKey logic. It is odd that at some places you are throwing exception and some methods return boolean.

Reply:
Updated the code based on the suggestion.